### PR TITLE
[ENH] `Plotly` and `Matplotlib` surface figures

### PR DESCRIPTION
**Pull request by NicolasGensollen**
Follow-up PR for #2902

This is a draft proposal for the implementation of a common interface for surface figures obtained with ``matplotlib`` and ``plotly``. For the moment, they are very basic and only implement `show` and `savefig` so you should be able to run:

```python
import matplotlib.pyplot as plt
from nilearn import plotting, surface, datasets
 
fsaverage = datasets.fetch_surf_fsaverage()
motor_images = datasets.fetch_neurovault_motor_task()
stat_img = motor_images.images[0]
texture = surface.vol_to_surf(stat_img, fsaverage.pial_right)
surf = surface.load_surf_mesh(fsaverage.pial_right)

for engine in ['plotly', 'matplotlib']:
    fig = plotting.plot_surf_stat_map(surf, texture, hemi='right',
                                      view='lateral', colorbar=True,
                                      threshold=1.2, bg_map=fsaverage['sulc_right'],
                                      title='Threshold + bg_map',
                                      engine=engine)
    fig.savefig(f"./plotting_with_{engine}.png")
    fig.show()
```

@jeromedockes this is only a first suggestion, we can iterate or close and start over. ;)

WIP...

See also #3028.

### CODE REVIEW

File: nilearn/plotting/displays.py [hunk 1]
```diff
@@ -2357,3 +2357,98 @@ def get_slicer(display_mode):
...
+        Output file path.
+
+    figure : Plotly figure instance or ``None``, optional
+        Figure.
+    """
+    def __init__(self, output_file=None, figure=None):
+        super().__init__(output_file=output_file, figure=figure)
+
+    @classmethod
+    def init_with_figure(cls, figure, output_file=None):
```
Comment (jeromedockes) [inline nilearn/plotting/displays.py:2413]: as we discussed not sure we need this one as it seems to duplicate `__init__`

File: nilearn/plotting/displays.py [hunk 2]
```diff
@@ -2357,3 +2357,98 @@ def get_slicer(display_mode):
...
+    """Implementation of a surface figure obtained with `plotly` engine.
+
+    Parameters
+    ----------
+    output_file : :obj:`str` or ``None``, optional
+        Output file path.
+
+    figure : Plotly figure instance or ``None``, optional
+        Figure.
+    """
```
Comment (jeromedockes) [inline nilearn/plotting/displays.py:2408]: maybe an `attributes` section where you point out to users that they can use `.figure` to access the underlying plotly figure for further customization and use of plotly functionality?

File: nilearn/plotting/displays.py [hunk 3]
```diff
@@ -2357,3 +2357,98 @@ def get_slicer(display_mode):
...
+            return self.figure
+
+    def savefig(self, output_file=None):
+        """Saves the figure to file.
+
+        Parameters
+        ----------
+        output_file : :obj:`str` or ``None``, optional
+            Path to output file.
+        """
```
Comment (jeromedockes) [inline nilearn/plotting/displays.py:2451]: do we need to check for the availability of kaleido here as well?

File: nilearn/plotting/displays.py (hunk 1 above)
Comment (NicolasGensollen) [reply nilearn/plotting/displays.py]: Thanks! Done in d2843cf

File: nilearn/plotting/displays.py (hunk 3 above)
Comment (NicolasGensollen) [reply nilearn/plotting/displays.py:2451]: Yes indeed! Thanks! Also done in d2843cf

**Review by jeromedockes — APPROVED** [review]
Thanks @NicolasGensollen I think it's good now!

File: nilearn/plotting/tests/test_surf_plotting.py [hunk 4]
```diff
@@ -70,6 +94,91 @@ def test_set_view_plot_surf_matplotlib(hemi, view, expected_view_matplotlib):
...
+                    reason=("Plotly and/or Ipython is not installed; "
+                            "required for this test."))
+@pytest.mark.parametrize("renderer", ['png', 'jpeg', 'svg'])
+def test_plotly_show(renderer):
+    ps = PlotlySurfaceFigure(go.Figure())
+    assert ps.output_file is None
+    assert ps.figure is not None
+    with mock.patch("IPython.display.display") as mock_display:
+        ps.show(renderer=renderer)
+    assert len(mock_display.call_args.args) == 1
```
Comment (bthirion) [inline nilearn/plotting/tests/test_surf_plotting.py:156]: these checks are slightly cryptic for non-experts. Briefly explain what is done here.

**Review by bthirion — COMMENTED** [review]
I think that the behavior should be illustrated in an example that does something with the generated figure.

**Comment by bthirion** [comment]
Thx !

**Comment by NicolasGensollen** [comment]
[Link](https://8804-1235740-gh.circle-artifacts.com/0/dev/auto_examples/01_plotting/plot_3d_map_to_surface_projection.html#sphx-glr-auto-examples-01-plotting-plot-3d-map-to-surface-projection-py) to the example.

**Review by bthirion — APPROVED** [review]
The example  is great. LGTM, thx !

**Comment by NicolasGensollen** [comment]
Thanks! Merging...

**Pull request merged**